internal/bus/bus_redis.go

```diff
@@ -19,71 +19,290 @@ import (
 	"crypto/sha256"
 	"encoding/base64"
 	"math/rand"
+	"sync"
 	"time"
 
+	"github.com/gammazero/deque"
 	"github.com/redis/go-redis/v9"
+	"go.uber.org/multierr"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
 	"google.golang.org/protobuf/proto"
+
+	"github.com/livekit/psrpc/internal/logger"
 )
 
 const lockExpiration = time.Second * 5
+const reconcilerRetryInterval = time.Second
 
 type redisMessageBus struct {
-	rc redis.UniversalClient
+	rc  redis.UniversalClient
+	ctx context.Context
+	ps  *redis.PubSub
+
+	mu     sync.Mutex
+	subs   map[string]*redisSubList
+	queues map[string]*redisSubList
+
+	wakeup          chan struct{}
+	ops             *deque.Deque[redisWriteOp]
+	dirtyChannels   *deque.Deque[string]
+	currentChannels map[string]struct{}
 }
 
 func NewRedisMessageBus(rc redis.UniversalClient) MessageBus {
-	return &redisMessageBus{
-		rc: rc,
+	ctx := context.Background()
+	r := &redisMessageBus{
+		rc:     rc,
+		ctx:    ctx,
+		ps:     rc.Subscribe(ctx),
+		subs:   map[string]*redisSubList{},
+		queues: map[string]*redisSubList{},
+
+		wakeup:          make(chan struct{}, 1),
+		ops:             deque.New[redisWriteOp](),
+		dirtyChannels:   deque.New[string](),
+		currentChannels: map[string]struct{}{},
 	}
+	go r.readWorker()
+	go r.writeWorker()
+	return r
 }
 
-func (r *redisMessageBus) Publish(ctx context.Context, channel string, msg proto.Message) error {
+func (r *redisMessageBus) Publish(_ context.Context, channel string, msg proto.Message) error {
 	b, err := serialize(msg)
 	if err != nil {
 		return err
 	}
 
-	return r.rc.Publish(ctx, channel, b).Err()
+	r.mu.Lock()
```

Comment:
is this just in case? IIRC we have not needed to lock Redis Publish previously.

Reply:
the lock protects the write queue

Reply:
sorry, what I meant to ask is if Publish should go parallel vs in a queue? Were there sync issues that warrants a lock + queue?

Reply:
it broke a unit test but we can add sleep calls... i tested parallel vs serialized publishes locally and there was no difference in throughput. if the redis call time was dominated by network latency it might be an issue since there is no way to batch writes

Reply:
we might need multiple write queues because blocking on publishes will break things but we need publishes for each channel to be ordered...

Reply:
yeah, latency is primarily what I was worried about. we could also use pipelines during publish.. but not sure if it's worth getting to that level of complexity.